File: solver.py
"""Equation solving logic, kept separate from the web layer so it's testable.

The public entry point is :func:`solve_equation`. Everything else is helpers.
"""

from sympy import Eq, Symbol, solveset
from sympy.core.sympify import SympifyError
from sympy.parsing.sympy_parser import (
    convert_xor,
    implicit_multiplication_application,
    parse_expr,
    standard_transformations,
)
from sympy.sets.sets import EmptySet

# Allow "2x" -> "2*x" and "x^2" -> "x**2" so input feels natural.
_TRANSFORMS = standard_transformations + (
    implicit_multiplication_application,
    convert_xor,
)


class SolveError(Exception):
    """Raised when an equation can't be parsed or solved sensibly."""


def _parse_side(text, side):
    try:
        return parse_expr(text, transformations=_TRANSFORMS, evaluate=True)
    except (SympifyError, SyntaxError, TypeError, ValueError):
        raise SolveError(f"Could not understand the {side} of the equation: '{text}'")


def _format_solution(value):
    """Render a SymPy value as a clean string, preferring exact then decimal."""
    exact = str(value)
    try:
        numeric = complex(value.evalf())
    except (TypeError, ValueError):
        return exact

    # Real result: also offer a rounded decimal if it differs from the exact form.
    if abs(numeric.imag) < 1e-12:
        real = numeric.real
        rounded = round(real, 6)
        decimal = ("%g" % rounded)
        if decimal != exact:
            return f"{exact} ≈ {decimal}" if "/" in exact or "sqrt" in exact or "**" in exact else exact
        return exact
    return exact


def solve_equation(equation, variable=None):
    """Solve ``equation`` (a string containing ``=``) for its unknown.

    Parameters
    ----------
    equation:
        Something like ``"10 + x = 15"``.
    variable:
        Optional name of the variable to solve for. If omitted, it is
        auto-detected (the equation must then have exactly one variable).

    Returns a dict with the variable name, the input echoed back, and a list
    of solution strings.
    """
    if "=" not in equation:
        raise SolveError("Equation must contain an '=' sign, e.g. 10 + x = 15.")
    if equation.count("=") > 1:
        raise SolveError("Equation should contain exactly one '=' sign.")

    left_text, right_text = equation.split("=")
    left = _parse_side(left_text.strip(), "left side")
    right = _parse_side(right_text.strip(), "right side")

    symbols = sorted((left - right).free_symbols, key=str)

    if variable:
        target = Symbol(variable)
        if target not in symbols:
            raise SolveError(
                f"The variable '{variable}' does not appear in the equation."
            )
    elif len(symbols) == 0:
        # No unknown: it's just a truth check, e.g. "2 + 2 = 4".
        holds = bool(Eq(left, right))
        verdict = "true" if holds else "false"
        return {
            "equation": equation,
            "variable": None,
            "solutions": [],
            "message": f"There is no variable to solve for. The statement is {verdict}.",
        }
    elif len(symbols) == 1:
        target = symbols[0]
    else:
        names = ", ".join(str(s) for s in symbols)
        raise SolveError(
            f"Found multiple variables ({names}). Tell me which one to solve for."
        )

    solution_set = solveset(Eq(left, right), target)

    if solution_set is EmptySet or solution_set == EmptySet:
        return {
            "equation": equation,
            "variable": str(target),
            "solutions": [],
            "message": f"No solution exists for {target}.",
        }

    try:
        values = list(solution_set)
    except TypeError:
        # Infinite / non-iterable solution set (e.g. identities like x = x).
        return {
            "equation": equation,
            "variable": str(target),
            "solutions": [],
            "message": f"{target} has infinitely many solutions: {solution_set}.",
        }

    solutions = [f"{target} = {_format_solution(v)}" for v in values]
    return {
        "equation": equation,
        "variable": str(target),
        "solutions": solutions,
        "message": None,
    }

File: static/solver.js
/* Equation-solving logic that runs entirely in the browser.
 *
 * This is the client-side port of the old Python `solver.py`. It leans on
 * nerdamer (loaded globally before this file) for the symbolic heavy lifting:
 * parsing, implicit multiplication (`2x`, `2(x+1)`), `^` powers, and solving.
 *
 * Public entry point: `solveEquation(equation, variable)`.
 * On bad input it throws a `SolveError` whose message is safe to show a user.
 */
(function (global) {
  "use strict";

  function SolveError(message) {
    this.name = "SolveError";
    this.message = message;
  }
  SolveError.prototype = Object.create(Error.prototype);
  SolveError.prototype.constructor = SolveError;

  // List the distinct variable names in an expression, sorted for stable output.
  function variablesOf(expr) {
    try {
      return nerdamer(expr).variables().slice().sort();
    } catch (e) {
      throw new SolveError("Could not understand the equation: '" + expr + "'.");
    }
  }

  // Render one solution: exact form, plus a rounded decimal when it isn't an
  // integer (so `10/3` shows as `10/3 ≈ 3.33333`, but `5` just shows `5`).
  function formatSolution(raw) {
    var exact;
    try {
      exact = nerdamer(raw).toString();
    } catch (e) {
      exact = String(raw);
    }

    var looksInexact = /[\/]|sqrt|\^|\*\*|root/i.test(exact);
    if (!looksInexact) return exact;

    var decimalText;
    try {
      decimalText = nerdamer(raw).evaluate().text("decimals");
    } catch (e) {
      return exact;
    }

    var num = Number(decimalText);
    if (!isFinite(num)) return exact; // complex/symbolic — leave exact as-is.

    // ~6 significant digits, trailing zeros trimmed, matching the old %g output.
    var rounded = Number(num.toPrecision(6));
    if (String(rounded) === exact) return exact;
    return exact + " ≈ " + rounded;
  }

  // Split nerdamer's "[a,b,...]" result into individual solution strings,
  // respecting any parentheses so commas inside an expression don't split it.
  function splitSolutions(solveResult) {
    if (solveResult && solveResult.symbol && solveResult.symbol.elements) {
      return solveResult.symbol.elements.map(function (e) {
        return e.toString();
      });
    }
    var text = solveResult.toString().trim().replace(/^\[/, "").replace(/\]$/, "");
    if (text === "") return [];
    var parts = [];
    var depth = 0;
    var current = "";
    for (var i = 0; i < text.length; i++) {
      var ch = text[i];
      if (ch === "(") depth++;
      if (ch === ")") depth--;
      if (ch === "," && depth === 0) {
        parts.push(current);
        current = "";
      } else {
        current += ch;
      }
    }
    if (current !== "") parts.push(current);
    return parts;
  }

  /**
   * Solve `equation` (a string containing `=`) for its unknown.
   *
   * @param {string} equation  e.g. "10 + x = 15"
   * @param {string} [variable] optional variable to solve for; auto-detected
   *                            when there's exactly one.
   * @returns {{equation:string, variable:?string, solutions:string[], message:?string}}
   */
  function solveEquation(equation, variable) {
    equation = (equation || "").trim();
    variable = (variable || "").trim();

    if (equation.indexOf("=") === -1) {
      throw new SolveError("Equation must contain an '=' sign, e.g. 10 + x = 15.");
    }
    if ((equation.match(/=/g) || []).length > 1) {
      throw new SolveError("Equation should contain exactly one '=' sign.");
    }

    var sides = equation.split("=");
    var left = sides[0].trim();
    var right = sides[1].trim();
    if (left === "" || right === "") {
      throw new SolveError("Both sides of the '=' need an expression.");
    }

    var diff = "(" + left + ")-(" + right + ")";
    var symbols = variablesOf(diff);

    var target;
    if (variable) {
      if (symbols.indexOf(variable) === -1) {
        throw new SolveError(
          "The variable '" + variable + "' does not appear in the equation."
        );
      }
      target = variable;
    } else if (symbols.length === 0) {
      // No unknown: it's just a truth check, e.g. "2 + 2 = 4".
      var holds;
      try {
        holds = nerdamer(diff).evaluate().toString() === "0";
      } catch (e) {
        throw new SolveError("Could not evaluate the equation.");
      }
      return {
        equation: equation,
        variable: null,
        solutions: [],
        message:
          "There is no variable to solve for. The statement is " +
          (holds ? "true" : "false") +
          ".",
      };
    } else if (symbols.length === 1) {
      target = symbols[0];
    } else {
      throw new SolveError(
        "Found multiple variables (" +
          symbols.join(", ") +
          "). Tell me which one to solve for."
      );
    }

    var result;
    try {
      result = nerdamer.solve(equation, target);
    } catch (e) {
      throw new SolveError("I couldn't solve that equation for " + target + ".");
    }

    var rawSolutions = splitSolutions(result);
    if (rawSolutions.length === 0) {
      return {
        equation: equation,
        variable: target,
        solutions: [],
        message: "No solution exists for " + target + ".",
      };
    }

    var solutions = rawSolutions.map(function (s) {
      return target + " = " + formatSolution(s);
    });

    return {
      equation: equation,
      variable: target,
      solutions: solutions,
      message: null,
    };
  }

  global.SolveError = SolveError;
  global.solveEquation = solveEquation;
})(typeof window !== "undefined" ? window : this);
